xcube/util/tilegrid.py

```diff
@@ -19,378 +19,351 @@
-                        ok = True
-                elif s_mode == MODE_LE:
-                    if s >= s_max >= s_max_min:
-                        ok = True
-                else:  # s_mode == MODE_EQ:
-                    if s_max == s:
-                        ok = True
-                    elif s_max > s:
-                        break
-                if ok:
-                    rec = s_max, ts, nt0, nl
-                    subdivisions.append(rec)
-
-    if not subdivisions:
-        return [(s, s, 1, 1)]
-
-    # maximize nl
-    subdivisions.sort(key=lambda r: r[3], reverse=True)
-    if ts_opt:
-        # minimize |ts - ts_opt|
-        subdivisions.sort(key=lambda r: abs(r[1] - ts_opt))
-    # minimize nt0
-    subdivisions.sort(key=lambda r: r[2])
-    # minimize s_max - s_min
-    subdivisions.sort(key=lambda r: r[0] - s)
-
-    return subdivisions
+            image_level_width = (image_level_width + 1) // 2
+            image_level_height = (image_level_height + 1) // 2
+            num_levels += 1
+
+        x_min, y_min = xy_min
+        extent = (x_min,
+                  y_min,
+                  x_min + image_width * xy_res,
+                  y_min + image_height * xy_res)
+
+        factor = 1 << (num_levels - 1)
+        max_resolution = xy_res * factor
+
+        super().__init__(
+            tile_size=(tile_width, tile_height),
+            num_level_0_tiles=(num_level_0_tiles_x, num_level_0_tiles_y),
+            crs=crs,
+            max_resolution=max_resolution,
+            extent=extent,
+            is_j_axis_up=is_j_axis_up,
+            max_level=num_levels - 1
+        )
+
+    def get_num_tiles(self, level: int) -> Tuple[int, int]:
+        image_width, image_height = self.get_image_size(level)
+        tile_width, tile_height = self.tile_size
+        return ((image_width + tile_width - 1) // tile_width,
+                (image_height + tile_height - 1) // tile_height)
+
+    def get_image_size(self, level: int) -> Tuple[int, int]:
+        factor = 1 << (self.max_level - level)
+        return self._image_width // factor, self._image_height // factor
+
+
+def tile_grid_to_ol_xyz_source_options(tile_grid: TileGrid, url: str) \
+        -> Optional[Dict[str, Any]]:
+    """
+    Convert :class:`TileGrid2` instance into options to be used with
+    ``ol.source.XYZ(options)`` of OpenLayers 4+.
+
+    See
+
+    * https://openlayers.org/en/latest/apidoc/module-ol_source_XYZ-XYZ.html
+    * https://openlayers.org/en/latest/apidoc/module-ol_tilegrid_TileGrid-TileGrid.html
+    * https://openlayers.org/en/latest/examples/xyz.html
+
+    :param tile_grid: tile grid
+    :param url: source url
+    :return: A JSON object that represents the ``options`` to be passed to
+        OpenLayers ``ol.source.XYZ(options)``.
+    """
+    if not tile_grid.crs.is_geographic:
+        return None
+
+    options = {
+        'url': url,
+        # 'projection': tile_grid.crs.srs
+        'projection': 'EPSG:4326'
```

Comment:
Is this correct?

Reply:
Yes, we ensure that the tile grid schema uses geographic coordinates. This is currently also what the viewer expect. Any other value will make it fail. There is still a lot of work for us...